ament_cmake_pytest/cmake/ament_add_pytest_test.cmake

```diff
@@ -0,0 +1,128 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Add a pytest test.
+#
+# :param testname: the name of the test
+# :type testname: string
+# :param path: the path to a file or folder where ``pytest`` should be invoked
+#   on
+# :type path: string
+# :param SKIP_TEST: if set mark the test as being skipped
+# :type SKIP_TEST: option
+# :param PYTHON_EXECUTABLE: absolute path to the executable used to run the test,
+#   default to the CMake variable with the same name returned by FindPythonInterp
+# :type PYTHON_EXECUTABLE: string
+# :param TIMEOUT: the test timeout in seconds,
+#   default defined by ``ament_add_test()``
+# :type TIMEOUT: integer
+# :param WORKING_DIRECTORY: the working directory for invoking the
+#   command in, default defined by ``ament_add_test()``
+# :type WORKING_DIRECTORY: string
+# :param ENV: list of env vars to set; listed as ``VAR=value``
+# :type ENV: list of strings
+# :param APPEND_ENV: list of env vars to append if already set, otherwise set;
+#   listed as ``VAR=value``
+# :type APPEND_ENV: list of strings
+# :param APPEND_LIBRARY_DIRS: list of library dirs to append to the appropriate
+#   OS specific env var, a la LD_LIBRARY_PATH
+# :type APPEND_LIBRARY_DIRS: list of strings
+#
+# @public
+#
+function(ament_add_pytest_test testname path)
+  cmake_parse_arguments(ARG
+    "SKIP_TEST"
+    "PYTHON_EXECUTABLE;TIMEOUT;WORKING_DIRECTORY"
+    "APPEND_ENV;APPEND_LIBRARY_DIRS;ENV"
+    ${ARGN})
+  if(ARG_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "ament_add_pytest_test() called with unused arguments: "
+      "${ARG_UNPARSED_ARGUMENTS}")
+  endif()
+
+  # check arguments
+  if(NOT IS_ABSOLUTE "${path}")
+    set(path "${CMAKE_CURRENT_SOURCE_DIR}/${path}")
+  endif()
+  # only check existence of path if it doesn't contain generator expressions
+  string(FIND "${path}" "$<" index)
+  if(index EQUAL -1 AND NOT EXISTS "${path}")
+    message(FATAL_ERROR
+      "ament_add_pytest_test() the path '${path}' does not exist")
+  endif()
+  if(NOT ARG_PYTHON_EXECUTABLE)
+    set(ARG_PYTHON_EXECUTABLE "${PYTHON_EXECUTABLE}")
+  endif()
+
+  # ensure pytest is available
+  ament_has_pytest(has_pytest QUIET PYTHON_EXECUTABLE "${ARG_PYTHON_EXECUTABLE}")
+  if(NOT has_pytest)
+    message(WARNING
+      "The Python module 'pytest' was not found, pytests can not be run (e.g. "
+      "on Ubuntu/Debian install the package 'python3-pytest')")
+    return()
+  endif()
+
+  set(result_file "${AMENT_TEST_RESULTS_DIR}/${PROJECT_NAME}/${testname}.xunit.xml")
+  set(cmd
+    "${ARG_PYTHON_EXECUTABLE}"
+    "-u"  # unbuffered stdout and stderr
+    "-m" "pytest"
+    "${path}"
+    # store last failed tests
+    "-o" "cache_dir=${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_pytest/${testname}/.cache"
```

Comment:
On Xenial I'm seeing an error with this option when testing rclpy. I installed pytest from apt `python3-pytest`.

```
1: usage: pytest.py [options] [file_or_dir] [file_or_dir] [...]
1: pytest.py: error: unrecognized arguments: -o cache_dir=/workspace/ros2_ws/build_isolated/rclpy/ament_cmake_pytest/rclpytests/.cache
```

Full output

```
test 1
    Start 1: rclpytests

1: Test command: /usr/bin/python3 "-u" "/workspace/ros2_ws/install_isolated/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/
workspace/ros2_ws/build_isolated/rclpy/test_results/rclpy/rclpytests.xunit.xml" "--output-file" "/workspace/ros2_ws/build_isolated/rclpy/
ament_cmake_pytest/rclpytests.txt" "--append-env" "AMENT_PREFIX_PATH=/workspace/ros2_ws/build_isolated/rclpy/ament_cmake_index" "LD_LIBRA
RY_PATH=/workspace/ros2_ws/build_isolated/rclpy" "--command" "/usr/bin/python3" "-u" "-m" "pytest" "/workspace/ros2_ws/src/ros2/rclpy/rcl
py/test" "-o" "cache_dir=/workspace/ros2_ws/build_isolated/rclpy/ament_cmake_pytest/rclpytests/.cache" "--junit-xml=/workspace/ros2_ws/bu
ild_isolated/rclpy/test_results/rclpy/rclpytests.xunit.xml" "--junit-prefix=rclpy"
1: Test timeout computed to be: 90
1: -- run_test.py: extra environment variables to append:
1:  - AMENT_PREFIX_PATH+=/workspace/ros2_ws/build_isolated/rclpy/ament_cmake_index
1:  - LD_LIBRARY_PATH+=/workspace/ros2_ws/build_isolated/rclpy
1: -- run_test.py: invoking following command in '/workspace/ros2_ws/build_isolated/rclpy':
1:  - /usr/bin/python3 -u -m pytest /workspace/ros2_ws/src/ros2/rclpy/rclpy/test -o cache_dir=/workspace/ros2_ws/build_isolated/rclpy/ame
nt_cmake_pytest/rclpytests/.cache --junit-xml=/workspace/ros2_ws/build_isolated/rclpy/test_results/rclpy/rclpytests.xunit.xml --junit-pre
fix=rclpy
1: usage: pytest.py [options] [file_or_dir] [file_or_dir] [...]
1: pytest.py: error: unrecognized arguments: -o cache_dir=/workspace/ros2_ws/build_isolated/rclpy/ament_cmake_pytest/rclpytests/.cache
1:   inifile: None
1:   rootdir: /workspace/ros2_ws
1: -- run_test.py: return code 2
1: -- run_test.py: generate result file '/workspace/ros2_ws/build_isolated/rclpy/test_results/rclpy/rclpytests.xunit.xml' with failed te$
t
1: -- run_test.py: verify result file '/workspace/ros2_ws/build_isolated/rclpy/test_results/rclpy/rclpytests.xunit.xml'
1/8 Test #1: rclpytests .......................***Failed    0.51 sec
```

Where is `-o cache_dir=` supposed be handled? I haven't found any references to `cache_dir` in cpython, unittest, or nose. I see some references to `cache_dir` in an ini in pytest [here](https://github.com/pytest-dev/pytest/blob/0a15edd5732578f6cc40e28770f0fc87dfa5ced9/testing/test_cache.py#L95).

Reply:
Looks like `cache_dir` was [added in version 3.2](https://docs.pytest.org/en/latest/customize.html#confval-cache_dir), but the version in apt is `2.8.7-4`. Installing via pip fixed the issue.

Reply:
wiki updated accordingly https://github.com/ros2/ros2/wiki/Linux-Development-Setup/_compare/192005729672b3367479a03efde46dd008d35f41...a60a4438fbcf7aa5747eb4a46e9acaedd7799d7e